lsp-mode.el

```diff
@@ -1416,9 +1423,72 @@ If WORKSPACE is not provided current workspace will be used."
     (concat (format "Content-Length: %d\r\n\r\n" (string-bytes body-with-newline))
             body-with-newline)))
 
+(cl-defstruct lsp--log-entry
+  (timestamp)
+  (process-time)
+  (type)
+  (method)
+  (id)
+  (body))
+
+(defun lsp--make-log-entry (method id body type &optional arg process-time)
+  "Create an outgoing log object from BODY with method METHOD and id ID.
+If ID is non-nil, then the body is assumed to be a notification.
+TYPE can either be 'incoming or 'outgoing"
+  (cl-assert (memq type '(incoming-req outgoing-req incoming-notif
+                                       outgoing-notif incoming-resp
+                                       outgoing-resp)))
+  (cl-check-type method string)
+  (when id (cl-check-type id number))
+  (make-lsp--log-entry
+   :timestamp (format-time-string "%I:%M:%S %p")
+   :process-time process-time
+   :method method
+   :id id
+   :type type
+   :body (or arg (plist-get body :params))))
+
+(defun lsp-save-logs (file)
+  (interactive "F")
+  (let ((trace (reverse (lsp--client-message-trace (lsp--workspace-client
+                                                    (nth 0 (lsp-workspaces))))))
```

Comment:
we may interactively ask which workspace to save if there are more than one workspaces active

Reply:
Is there a way to map workspaces to a user-selectable string?

Reply:
yeah - lsp--workspace-print

Reply:
Done, thanks.